### ISSUE TITLE
Logging from an Email Service does not generate a log

### ISSUE THREAD

**Issue by arbokrad** [jongpie/NebulaLogger#155]
(continued here after discord troubleshooting, thanks @jongpie !_)

Nebula Version: 4.3
Edition: Managed Package

Attempting to log from an Email Service class (a class implementing **Messaging.InboundEmailHandler**) does not appear to generate logs, while the same user running anonymous apex to generate a log does.

Subscribing to the Nebula__LogEntryEvent__e event in workbench shows an event for the anonymous apex execution, but **not for the email service execution**, despite other logs indicating that it attempts to create an event:

```
13:53:13.0 (440689677)|USER_DEBUG|[12]|DEBUG|nebula buffer: 1
13:53:13.0 (440702540)|STATEMENT_EXECUTE|[13]
13:53:13.0 (440717117)|METHOD_ENTRY|[13]|01p1700000HMH54|Nebula.Logger.saveLog()
13:53:13.0 (450641438)|DML_BEGIN|[4765]|Op:Insert|Type:Nebula__LogEntryEvent__e|Rows:1
13:53:13.0 (460136830)|DML_END|[4765]
13:53:13.0 (470101036)|ENTERING_MANAGED_PKG|Nebula
13:53:13.0 (470285226)|METHOD_EXIT|[13]|01p1700000HMH54|Nebula.Logger.saveLog()
13:53:13.0 (470303279)|STATEMENT_EXECUTE|[14]
13:53:13.0 (470317048)|METHOD_ENTRY|[14]|01p1700000HMH54|Nebula.Logger.getBufferSize()
13:53:13.0 (470329493)|ENTERING_MANAGED_PKG|Nebula
13:53:13.0 (470410130)|METHOD_EXIT|[14]|01p1700000HMH54|Nebula.Logger.getBufferSize()
```

The Automated Process user also does not generate any kind of debug logs when executing from the email service, but does with the similar code under anonymous apex.

Sample code
========

Anonymous apex (generates a log as-expected):
```
try {
	String s = null;
	s.length();
} catch( Exception ex ) {
	Nebula.Logger.error(ex.getMessage(), ex);
} finally  {
	Nebula.Logger.saveLog();
}
```

Rough email handler code (no logs generated):
```
global class NebulaEmailHandler implements Messaging.InboundEmailHandler {
	global Messaging.InboundEmailResult handleInboundEmail(Messaging.InboundEmail email, Messaging.InboundEnvelope env) {

		Messaging.InboundEmailResult result = new Messaging.InboundEmailResult();
		try {
			String s = null;
			s.length();
		} catch (Exception ex) {
			Nebula.Logger.error(ex.getMessage(), ex);
		} finally {
			result.success = true;
			System.debug('nebula buffer: ' + Nebula.Logger.getBufferSize());
			Nebula.Logger.saveLog();
			System.debug('nebula buffer: ' + Nebula.Logger.getBufferSize());
		}

		return result;
	}
}
```

**Comment by jongpie** [comment]
@arbokrad thanks again for logging this issue! I tried to reproduce this issue, but everything seems to be working for me (see PR #157):

- I've created a branch with an example email service class, [ExampleInboundEmailHandler](https://github.com/jongpie/NebulaLogger/blob/feature/more-tests/extra-tests/extra-metadata/classes/ExampleInboundEmailHandler.cls) - this is more or less the same example implementation you provided for `Messaging.InboundEmailHandler`
- I also created a test class, [Logger_Tests_InboundEmailHandler](https://github.com/jongpie/NebulaLogger/blob/feature/more-tests/extra-tests/Logger_Tests_InboundEmailHandler.cls) - it calls the email service class to verify it creates a log, and the test is passing for me

I'm still stumped on why it's not working for you in this particular scenario - can you see any major differences in my tests compared to what's in your org?

**Comment by arbokrad** [comment]
(following more work on this today)

It appears that behavior between apex invoking the class/method and the actual email service context may be different. Under direct apex, all works as expected (the proper debug logs are generated, the proper Nebula logs are generated)

**Reproduction steps for manual testing**
1. Create a new dev org
2. Install nebula (I tried both managed and unlocked versions)
3. set up org-default Logger Settings
4. Assign **Logger: Admin** permset to the dev org admin user
5. Create a new email handler class (I used the **ExampleInboundEmailHandler**, from the prior comment)
6. Configure the email service to use the class, create a new email address and save it for later.  (Note: remove any emails from the "accept email from" field, as this restricts who can email this endpoint and for testing you may need other users)
7. Setup debug logs to trace both the automated process and the admin user (whoever is configured to run the email service)
8. Send an email to the email address above (the contents are irrelevant as long as it goes to that address)
9. Note that the debug logs show _only_ for the user running the email service and not the expected automated process, and no Log__c record is generated

Other things we tried:
1. Changing the error settings on the email service (from discard to bounce, no effect)
2. Setting error emails to be routed to a different address (hoping to get a log, no effect)
3.

**Comment by jongpie** [comment]
@arbokrad I think I've finally figured out the root cause of this issue! Short answer: 1 null string and 1 assumption have caused havoc, but it should hopefully be a quick fix that I can make in the next week or two.

Long answer: It appears that the issue is with the String being used for `Logger.getTransactionId()` - behind the scenes, the transaction ID is set based on `System.Request.getCurrent().getRequestId()` ([see docs](https://developer.salesforce.com/docs/atlas.en-us.apexref.meta/apexref/apex_class_System_Request.htm)), which appears to be null when running in the context of an `InboundEmailHandler`.

![image](https://user-images.githubusercontent.com/1267157/122651178-e7378700-d0eb-11eb-9d66-d7a46da5b6a1.png)

This isn't something we thought could happen - I was under the impression that Salesforce's method `getRequestId()` would always return a value. But since it's null, the platform events are failing to insert because [the transaction ID is a required field](https://github.com/jongpie/NebulaLogger/blob/main/nebula-logger/main/logger-engine/objects/LogEntryEvent__e/fields/TransactionId__c.field-meta.xml#L10) on the platform event object. 

I think the solution will be to change `Logger` to use a secondary unique String for the transaction ID in the event that `System.Request.getCurrent().getRequestId()` is null. In older versions of Logger, I was actually using [my UUID class](https://github.com/jongpie/ApexUuid/) as the unique transaction ID - I'll probably re-integrate some of that code, and use a UUID as a fallback unique value for the transaction ID.

**Comment by jongpie** [comment]
@arbokrad I went ahead and merged PR #170 to (finally) fix this bug - this will be included in the next release, v4.5.0, via PR #165. I'm hoping to have the next release finished within the next few days

**Comment by jongpie** [comment]
@arbokrad I've released v4.5.0 of the packages! You can install the latest managed package version [with this link](https://login.salesforce.com/packaging/installPackage.apexp?p0=04t5Y0000027FMhQAM).
